In [97]:
import numpy as np
import math

In [98]:
patient_array = np.array([[76,126,38.0], [74,120,38.0], [74,118,37.5], [78,136,37.0]])

print(patient_array)

[[ 76.  126.   38. ]
 [ 74.  120.   38. ]
 [ 74.  118.   37.5]
 [ 78.  136.   37. ]]


In [99]:
avg = np.mean(patient_array, axis = 0)              # Average of all features
print(avg)

[ 75.5   125.     37.625]


In [100]:
distance = []                                   # Stores the distances from average point

for i,ele in enumerate(patient_array):          # Subtract vectors and find the distance
    temp = (patient_array[i] - avg) ** 2             
    dist = math.sqrt(temp.sum())
    distance.append(dist)

print(distance)
                                                  # Adding 1 to index since python index starts from 0              
print("The patient farthest from the rest is: " + str(distance.index(max(distance)) + 1))


[1.1792476415070754, 5.23360535386458, 7.160001745809843, 11.297815054248321]
The patient farthest from the rest is: 4


In [101]:
n = patient_array.shape[0]                  # taking number of rows
print(n)

dis = []
i=0
j=0

while i < n-1:                                 # To calculate distance from each and every element and store in 'dis' list                 
    while j < n-1:
        temp1 = (patient_array[i] - patient_array[j+1]) ** 2
        dist = math.sqrt(temp1.sum())
        dis.append([dist,i,j+1])                # 'dis' contains list of [distance of 2 element, 1st element, 2nd element]
        print(i,j+1)                            # patients whoes distance have been calculated in this iteration
        j += 1
    i += 1
    j = i

print(dis)

dis = np.array(dis)

minimun = dis[0][0]
                                                        # Find the minimum in 'dis 
for i,ele in enumerate(dis):
    if minimun > dis[i,0]:
        minimun = dis[i,0]
    


n = np.argwhere(dis == minimun)[0,0]                    # return the index of the minimun element 

print(n)
                                                            # Adding 1 to index since python index starts from 0
print("Nearest patients are: " +  str(int(dis[n,1]) + 1) + " and " + str(int(dis[n,2]) + 1))

4
0 1
0 2
0 3
1 2
1 3
2 3
[[6.324555320336759, 0, 1], [8.261355820929152, 0, 2], [10.246950765959598, 0, 3], [2.0615528128088303, 1, 2], [16.522711641858304, 1, 3], [18.445866745696716, 2, 3]]
3
Nearest patients are: 2 and 3


In [102]:
pat_5 = [100,128,37.2]                                  # declare new patient

di = []                                                 # 'di' list to store distance of all patients from 5th patients

for i,ele in enumerate(patient_array):                  # Store elements in 'di'
    t = (pat_5 - patient_array[i]) ** 2
    d = math.sqrt(t.sum())
    di.append(d)

print(di)

print(min(di))          
                                                # Adding 1 to index since python index starts from 0
print("The nearest patient to Patient 5 is:" + " " + str(di.index(min(di))+1))

[24.096472770926454, 27.21470190907848, 27.8583919133894, 23.410254163507066]
23.410254163507066
The nearest patient to Patient 5 is: 4


1.1 In this program the time complexity of finding farthest patient is O(n) where n is the row elements of patient array
 
1.2 The time complexity of finding 2 close patients is O($n^2$)

2 We could use it to find different patients of severe condition and seperate then. If there was a feature like "Insurance", we could use it to segregate patients who have and have not paid insurace, thus saving time in managing a hospital.  

